### ISSUE TITLE
could not write data to host: Connection refused in write() while using Shoutcast stream

### ISSUE THREAD

**Issue by itcrowdsource** [LibreTime/libretime#686]
**Describe the bug**
The latest release of LibreTime is having an issue with Shoutcast DNAS services. If Shoutcast is being configured, LibreTime throws an error "could not write data to host: Connection refused in write()". Other (older) instances of LibreTime are running fine within our environment. We have also reinstalled the entire server to eliminate any unknown issues from the previous install. But the problem still persists. 
I noticed there is version difference of **rabbitmq** between the older and the newer instances. The old LibreTime instances are running 3.7.5-1 and the new one 3.7.10-1. Tried to downgrade the **rabbitmq** version, but that breaks the installer.

SYSLOG
```
Jan 11 15:36:51 stream airtime-liquidsoap[23952]: 2019-01-11 15:36:51,402 [pyponotify] [INFO ]  #################################################
Jan 11 15:36:51 stream airtime-liquidsoap[23952]: 2019-01-11 15:36:51,402 [pyponotify] [INFO ]  # Calling server to update liquidsoap status    #
Jan 11 15:36:51 stream airtime-liquidsoap[23952]: 2019-01-11 15:36:51,403 [pyponotify] [INFO ]  #################################################
Jan 11 15:36:51 stream airtime-liquidsoap[23952]: 2019-01-11 15:36:51,403 [pyponotify] [INFO ]  msg = could not write data to host: Connection refused in write()
Jan 11 15:36:51 stream airtime-liquidsoap[23952]: 2019-01-11 15:36:51,491 [pyponotify] [INFO ]  Response: null
Jan 11 15:36:51 stream airtime-liquidsoap[23952]: #########################################
Jan 11 15:36:51 stream airtime-liquidsoap[23952]: #           *** pypo  ***               #
Jan 11 15:36:51 stream airtime-liquidsoap[23952]: #     pypo notification gateway         #
Jan 11 15:36:51 stream airtime-liquidsoap[23952]: #########################################
```

```
2019/01/11 15:37:57 [shoutcast_stream_1:3] Will try again in 5.00 sec.
2019/01/11 15:38:03 [shoutcast_stream_1:3] Connecting mount / for source@xxxxxxx...
2019/01/11 15:38:03 [shoutcast_stream_1:2] Connection failed: could not write data to host: Connection refused in write()
2019/01/11 15:38:03 [lang:3] timeout --signal=KILL 45 pyponotify --error='could not write data to host: Connection refused in write()' --stream-id=1 --time=1547220063.09 &
```

ls_script.log
**To Reproduce**
1. Clean install of LibreTime with default settings
2. Installed and started Shoutcast DNAS
3. Configured LibreTime stream to use Shoutcast service (Custom / 3rd Party Streaming)
4. After saving Shoutcast stream settings the following error appears : "could not write data to host: Connection refused in write()"

**Expected behavior**
The Shoutcast stream should work

**LibreTime version:**
Version from the upgrade popup if you can reach it.

**Installation method:**
 - OS: Ubuntu
 - OS Version: 16.04 LTS
 - Method: ./install
 - Details: git clone https://github.com/LibreTime/libretime.git

**Screenshots**
![image](https://user-images.githubusercontent.com/20809792/51043746-bf10d900-15bf-11e9-80aa-21c062fd455c.png)

**Client (please complete the following information if applicable):**
 - OS: Windows 10
 - Browser Edge, IE, Chrome, FireFox

**Comment by itcrowdsource** [comment]
I also noticed that this new version doesn't playout properly. The player on the landing page and the listen button aren't playing out the stream although the Icecast stream is up and running.

**Comment by hairmare** [comment]
I recently tested with liquidsoap 1.3.5 and didn't experience any issues. Can you post the output of `liquidsoap --version` so we are sure what version you are using?
Please also post the start of you liquisoap logs, it has some information on how liquidsoap is compiled and what versions of the dependencies it is using.
How exactly did you install liquidsoap? Are you using an opam install or packages?

If you want to downgrade liquidsoap you don't need to rerun the installer, just restart the playout services and you should be good to go.

The `could not write data to host: Connection refused in write()` error seems to indicate that the shoutcast server isn't reachable/running. Liquidsoap is probably shutting down after this error so it isn't sending any thing to Icecast anymore either. Are you really using both Icecast and SHOUTcast?

**Comment by itcrowdsource** [comment]
Sorry, I meant there is a difference in the version of **rabbitmq**

**Comment by itcrowdsource** [comment]
This is the liquidsoap log while restarting the service:
```
2019/01/12 11:15:29 [lang:3] timeout --signal=KILL 45 pyponotify --error='could not write data to host: Connection refused in write()' --stream-id=1 --time=1547291273.43 &
2019/01/12 11:15:29 [shoutcast_stream_1:3] Will try again in 5.00 sec.
2019/01/12 11:15:35 [shoutcast_stream_1:3] Connecting mount / for source@<URL>...
2019/01/12 11:15:35 [shoutcast_stream_1:2] Connection failed: could not write data to host: Connection refused in write()
2019/01/12 11:15:35 [lang:3] timeout --signal=KILL 45 pyponotify --error='could not write data to host: Connection refused in write()' --stream-id=1 --time=1547291273.43 &
2019/01/12 11:15:35 [shoutcast_stream_1:3] Will try again in 5.00 sec.
2019/01/12 11:15:41 [shoutcast_stream_1:3] Connecting mount / for source@<URL>...
2019/01/12 11:15:41 [shoutcast_stream_1:2] Connection failed: could not write data to host: Connection refused in write()
2019/01/12 11:15:41 [lang:3] timeout --signal=KILL 45 pyponotify --error='could not write data to host: Connection refused in write()' --stream-id=1 --time=1547291273.43 &
2019/01/12 11:15:41 [shoutcast_stream_1:3] Will try again in 5.00 sec.
2019/01/12 11:15:47 [shoutcast_stream_1:3] Connecting mount / for source@<URL>...
2019/01/12 11:15:47 [shoutcast_stream_1:2] Connection failed: could not write data to host: Connection refused in write()
2019/01/12 11:15:47 [lang:3] timeout --signal=KILL 45 pyponotify --error='could not write data to host: Connection refused in write()' --stream-id=1 --time=1547291273.43 &
2019/01/12 11:15:47 [shoutcast_stream_1:3] Will try again in 5.00 sec.
2019/01/12 11:15:53 [shoutcast_stream_1:3] Connecting mount / for source@<URL>...
2019/01/12 11:15:53 [shoutcast_stream_1:2] Connection failed: could not write data to host: Connection refused in write()
2019/01/12 11:15:53 [lang:3] timeout --signal=KILL 45 pyponotify --error='could not write data to host: Connection refused in write()' --stream-id=1 --time=1547291273.43 &
2019/01/12 11:15:53 [shoutcast_stream_1:3] Will try again in 5.00 sec.
2019/01/12 11:15:54 [server:3] New client: localhost.
2019/01/12 11:15:54 [lang:3] dynamic_source.get_id
2019/01/12 11:15:54 [server:3] Client localhost disconnected.
2019/01/12 11:15:59 [shoutcast_stream_1:3] Connecting mount / for source@<URL>...
2019/01/12 11:15:59 [shoutcast_stream_1:2] Connection failed: could not write data to host: Connection refused in write()
2019/01/12 11:15:59 [lang:3] timeout --signal=KILL 45 pyponotify --error='could not write data to host: Connection refused in write()' --stream-id=1 --time=1547291273.43 &
2019/01/12 11:15:59 [shoutcast_stream_1:3] Will try again in 5.00 sec.
2019/01/12 11:16:05 [shoutcast_stream_1:3] Connecting mount / for source@<URL>...
2019/01/12 11:16:05 [shoutcast_stream_1:2] Connection failed: could not write data to host: Connection refused in write()
2019/01/12 11:16:05 [lang:3] timeout --signal=KILL 45 pyponotify --error='could not write data to host: Connection refused in write()' --stream-id=1 --time=1547291273.43 &
2019/01/12 11:16:05 [shoutcast_stream_1:3] Will try again in 5.00 sec.
2019/01/12 11:16:05 [main:3] Shutdown started!
2019/01/12 11:16:05 [main:3] Waiting for threads to terminate...
2019/01/12 11:16:05 [stream:3] Closing connection...
2019/01/12 11:16:05 [lang:3] timeout --signal=KILL 45 pyponotify --source-name=live_dj --source-status=false &
2019/01/12 11:16:05 [harbor:3] Removing mountpoint '/show' on port 8002
2019/01/12 11:16:05 [harbor:3] Nothing more on port 8002: closing sockets.
2019/01/12 11:16:05 [lang:3] timeout --signal=KILL 45 pyponotify --source-name=master_dj --source-status=false &
2019/01/12 11:16:05 [harbor:3] Removing mountpoint '/master' on port 8001
2019/01/12 11:16:05 [harbor:3] Nothing more on port 8001: closing sockets.
2019/01/12 11:16:05 [clock.wallclock_main:3] Streaming loop stopped.
2019/01/12 11:16:05 [threads:3] Thread "wallclock_main" terminated (1 remaining).
2019/01/12 11:16:05 [threads:3] Thread "http polling" terminated (0 remaining).
2019/01/12 11:16:05 [main:3] Cleaning downloaded files...
2019/01/12 11:16:05 [main:3] Freeing memory...
2019/01/12 11:16:05 >>> LOG END
2019/01/12 11:16:06 >>> LOG START
2019/01/12 11:16:06 [protocols.external:3] Found "/usr/bin/wget".
2019/01/12 11:16:06 [main:3] Liquidsoap 1.1.1
2019/01/12 11:16:06 [main:3] Using: graphics=[distributed with Ocaml] pcre=7.0.4 dtools=0.3.1 duppy=0.5.1 duppy.syntax=0.5.1 cry=0.2.2 mm=0.2.1 xmlplaylist=0.1.3 lastfm=0.3.0 ogg=0.4.5 vorbis=0.6.1 opus=0.1.0 speex=0.2.0 mad=0.4.4 flac=0.1.1 flac.ogg=0.1.1 dynlink=[distributed with Ocaml] lame=0.3.2 shine=0.2.0 gstreamer=0.2.0 frei0r=0.1.0 voaacenc=0.1.0 theora=0.3.0 schroedinger=0.1.0 gavl=0.1.5 bjack=0.1.4 alsa=0.2.1 ao=0.2.0 samplerate=0.1.2 taglib=0.3.1 magic=0.7.3 camomile=0.8.4 inotify=1.0 faad=0.3.2 soundtouch=0.1.7 portaudio=0.2.0 pulseaudio=0.1.2 ladspa=0.1.4 dssi=0.1.1 sdl=0.9.1 camlimages=4.2.0 lo=0.1.0 yojson=1.2.3 gd=1.0a5
2019/01/12 11:16:06 [dynamic.loader:3] Could not find dynamic module for fdkaac encoder.
2019/01/12 11:16:06 [dynamic.loader:3] Could not find dynamic module for aacplus encoder.
2019/01/12 11:16:06 [dynamic.loader:2] Loaded plugin file /usr/lib/liquidsoap/1.1.1/plugins/lame.cmxs.
2019/01/12 11:16:06 [dynamic.loader:2] Loaded plugin file /usr/lib/liquidsoap/1.1.1/plugins/cry.cmxs.
2019/01/12 11:16:06 [dynamic.loader:2] Loaded plugin file /usr/lib/liquidsoap/1.1.1/plugins/ogg.cmxs.
2019/01/12 11:16:06 [dynamic.loader:2] Loaded plugin file /usr/lib/liquidsoap/1.1.1/plugins/ao.cmxs.
2019/01/12 11:16:06 [dynamic.loader:2] Loaded plugin file /usr/lib/liquidsoap/1.1.1/plugins/voaacenc.cmxs.
2019/01/12 11:16:06 [dynamic.loader:2] Loaded plugin file /usr/lib/liquidsoap/1.1.1/plugins/faad.cmxs.
2019/01/12 11:16:06 [dynamic.loader:2] Loaded plugin file /usr/lib/liquidsoap/1.1.1/plugins/flac.cmxs.
2019/01/12 11:16:06 [dynamic.loader:2] Loaded plugin file /usr/lib/liquidsoap/1.1.1/plugins/mad.cmxs.
2019/01/12 11:16:06 [dynamic.loader:2] Loaded plugin file /usr/lib/liquidsoap/1.1.1/plugins/flac_ogg.cmxs.
2019/01/12 11:16:06 [dynamic.loader:2] Loaded plugin file /usr/lib/liquidsoap/1.1.1/plugins/portaudio.cmxs.
2019/01/12 11:16:06 [dynamic.loader:2] Loaded plugin file /usr/lib/liquidsoap/1.1.1/plugins/pulseaudio.cmxs.
2019/01/12 11:16:06 [dynamic.loader:2] Loaded plugin file /usr/lib/liquidsoap/1.1.1/plugins/vorbis.cmxs.
2019/01/12 11:16:06 [dynamic.loader:2] Loaded plugin file /usr/lib/liquidsoap/1.1.1/plugins/alsa.cmxs.
2019/01/12 11:16:06 [dynamic.loader:2] Loaded plugin file /usr/lib/liquidsoap/1.1.1/plugins/taglib.cmxs.
2019/01/12 11:16:06 [frame:3] Using 44100Hz audio, 25Hz video, 44100Hz master.
2019/01/12 11:16:06 [frame:3] Frame size must be a multiple of 1764 ticks = 1764 audio samples = 1 video samples.
2019/01/12 11:16:06 [frame:3] Targetting 'frame.duration': 0.04s = 1764 audio samples = 1764 ticks.
2019/01/12 11:16:06 [frame:3] Frames last 0.04s = 1764 audio samples = 1 video samples = 1764 ticks.
2019/01/12 11:16:06 [lang:3] timeout --signal=KILL 45 pyponotify --liquidsoap-started &
2019/01/12 11:16:06 [threads:3] Created thread "generic queue #1".
2019/01/12 11:16:06 [threads:3] Created thread "generic queue #2".
2019/01/12 11:16:06 [threads:3] Created thread "non-blocking queue #1".
2019/01/12 11:16:06 [threads:3] Created thread "non-blocking queue #2".
2019/01/12 11:16:06 [harbor:3] Adding mountpoint '/master' on port 8001
2019/01/12 11:16:06 [harbor:3] Adding mountpoint '/show' on port 8002
2019/01/12 11:16:06 [threads:3] Created thread "http polling" (1 total).
2019/01/12 11:16:06 [shoutcast_stream_1:3] Connecting mount / for source@<URL>...
2019/01/12 11:16:06 [shoutcast_stream_1:2] Connection failed: could not write data to host: Connection refused in write()
2019/01/12 11:16:06 [lang:3] timeout --signal=KILL 45 pyponotify --error='could not write data to host: Connection refused in write()' --stream-id=1 --time=1547291766.79 &
2019/01/12 11:16:06 [shoutcast_stream_1:3] Will try again in 5.00 sec.
2019/01/12 11:16:06 [stream:3] Connecting mount stream for source@<URL>...
2019/01/12 11:16:06 [stream:3] Connection setup was successful.
2019/01/12 11:16:06 [lang:3] timeout --signal=KILL 45 pyponotify --connect --stream-id=3 --time=1547291766.79 &
2019/01/12 11:16:06 [threads:3] Created thread "wallclock_main" (2 total).
2019/01/12 11:16:06 [clock.wallclock_main:3] Streaming loop starts, synchronized with wallclock.
2019/01/12 11:16:06 [dummy:3] Source failed (no more tracks) stopping output...
2019/01/12 11:16:06 [dummy(dot)1:3] Source failed (no more tracks) stopping output...
2019/01/12 11:16:06 [dummy(dot)2:3] Source failed (no more tracks) stopping output...
2019/01/12 11:16:06 [dummy(dot)3:3] Source failed (no more tracks) stopping output...
2019/01/12 11:16:06 [dummy(dot)4:3] Source failed (no more tracks) stopping output...
2019/01/12 11:16:06 [map_metadata_5856:3] Inserting missing metadata.
2019/01/12 11:16:06 [master_show_schedule_noise_switch:3] Switch to show_schedule_noise_switch.
2019/01/12 11:16:06 [show_schedule_noise_switch:3] Switch to schedule_noise_switch.
2019/01/12 11:16:06 [schedule_noise_switch:3] Switch to map_metadata_5856.
2019/01/12 11:16:07 [server:3] New client: localhost.
2019/01/12 11:16:07 [lang:3] streams.scheduled_play_start
2019/01/12 11:16:07 [lang:3] streams.live_dj_stop
2019/01/12 11:16:07 [lang:3] streams.master_dj_stop
2019/01/12 11:16:07 [lang:3] vars.stream_metadata_type
2019/01/12 11:16:07 [lang:3] vars.station_name
2019/01/12 11:16:07 [lang:3] vars.default_dj_fade
2019/01/12 11:16:07 [server:3] Client localhost disconnected.
2019/01/12 11:16:07 [server:3] New client: localhost.
2019/01/12 11:16:07 [lang:3] queues.s8_skip
2019/01/12 11:16:07 [lang:3] queues.s8_skip
2019/01/12 11:16:07 [lang:3] queues.s8_skip
2019/01/12 11:16:07 [lang:3] queues.s8_skip
2019/01/12 11:16:07 [lang:3] queues.s8_skip
2019/01/12 11:16:07 [server:3] Client localhost disconnected.
2019/01/12 11:16:07 [schedule_noise_switch:3] Switch to map_metadata_5849 with transition.
2019/01/12 11:16:07 [lang:3] transition called...
2019/01/12 11:16:07 [switch_5847:3] Switch to insert_metadata_5833.
2019/01/12 11:16:07 [cross_5781:3] No next track ready yet.
2019/01/12 11:16:07 [cross_5766:3] No next track ready yet.
2019/01/12 11:16:07 [cross_5751:3] No next track ready yet.
2019/01/12 11:16:07 [cross_5736:3] No next track ready yet.
2019/01/12 11:16:07 [cross_5721:3] No next track ready yet.
2019/01/12 11:16:07 [dummy(dot)4:3] Source failed (no more tracks) stopping output...
2019/01/12 11:16:07 [server:3] New client: localhost.
2019/01/12 11:16:07 [lang:3] dynamic_source.get_id
2019/01/12 11:16:07 [server:3] Client localhost disconnected.
2019/01/12 11:16:07 [schedule_noise_switch:3] Switch to map_metadata_5856 with transition.
2019/01/12 11:16:07 [lang:3] transition called...
2019/01/12 11:16:12 [shoutcast_stream_1:3] Connecting mount / for source@<URL>...
2019/01/12 11:16:12 [shoutcast_stream_1:2] Connection failed: could not write data to host: Connection refused in write()
2019/01/12 11:16:12 [lang:3] timeout --signal=KILL 45 pyponotify --error='could not write data to host: Connection refused in write()' --stream-id=1 --time=1547291766.79 &
2019/01/12 11:16:12 [shoutcast_stream_1:3] Will try again in 5.00 sec.
2019/01/12 11:16:18 [shoutcast_stream_1:3] Connecting mount / for source@<URL>...
2019/01/12 11:16:18 [shoutcast_stream_1:2] Connection failed: could not write data to host: Connection refused in write()
2019/01/12 11:16:18 [lang:3] timeout --signal=KILL 45 pyponotify --error='could not write data to host: Connection refused in write()' --stream-id=1 --time=1547291766.79 &
2019/01/12 11:16:18 [shoutcast_stream_1:3] Will try again in 5.00 sec.
```

**Comment by Robbt** [comment]
So it seems like your Shoutcast server might be blocking the connection for some reason. Are you running your own shoutcast server if so can you check the logs of it to see if it shows a connection attempt. It could also be a network related error. 
Someone previously had this same error and looking on the Internet people had it before when dealing with AAC codec and also when their service provider hadn't been paid so that leads me to believe that the issue is the connection.

It also seems like something else is probably broken if you say playout isn't working with Icecast either and perhaps that is the problem. I don't have a shoutcast server to test it with and unless there is a free one somewhere there isn't much I can do other than guess at what might be the problem.

I'd basically scan all of the logs in /var/log/airtime and see if you notice any other errors with pypo or celery or zend-php that might provide more clues as to what might be not working. Hope we can figure this out but I don't have enough information to do anything other than ask for more information at this point.

**Comment by itcrowdsource** [comment]
I finally found the cause of this issue after testing access to the Shoutcast port on the host itself. I was able to reach the Shoutcast management page so initially assumed that Shoutcast was running properly. However I discovered that the Shoutcast streaming service wasn't reachable.
Somehow the hosting provider had changed the default OS deployment image. The new OS image contains a pre-installed/pre-configured DNS service which resolved the FQDN in an incorrect way (to 127.0.0.1 and 127.0.1.1). So various network bindings that were configured to use the external network address (including Shoutcast) didn't work properly.
